server/appender.go

```diff
@@ -0,0 +1,572 @@
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/metadata"
+	"github.com/prometheus/prometheus/model/textparse"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/model/value"
+	"github.com/prometheus/prometheus/scrape"
+	"github.com/prometheus/prometheus/storage"
+)
+
+var errNameLabelMandatory = fmt.Errorf("missing metric name (%s label)", labels.MetricName)
+
+type appendErrors struct {
+	numOutOfOrder         int
+	numDuplicates         int
+	numOutOfBounds        int
+	numExemplarOutOfOrder int
+}
+
+// appendFn an scrapecache are copied from prometheus project. Ideally it should've been used from the public API.
+// The public API allows using the ScrapeManager only. We can later use scrape.ScrapeManager but for
+// now we are only using the bits necessary for the project.
+func appendFn(app storage.Appender, logger log.Logger, cache *scrapeCache, b []byte, contentType string, ts time.Time, conf *configuration) error {
+	p, err := textparse.New(b, contentType, true)
+	if err != nil {
+		_ = level.Debug(logger).Log(
+			"msg", "Invalid content type on scrape, using prometheus parser as fallback.",
+			"content_type", contentType,
+			"err", err,
+		)
+	}
+
+	var (
+		defTime         = timestamp.FromTime(ts)
+		appErrs         = appendErrors{}
+		sampleLimitErr  error
+		bucketLimitErr  error
+		lset            labels.Labels     // escapes to heap so hoisted out of loop
+		e               exemplar.Exemplar // escapes to heap so hoisted out of loop
+		meta            metadata.Metadata
+		metadataChanged bool
+	)
+
+	// updateMetadata updates the current iteration's metadata object and the
+	// metadataChanged value if we have metadata in the scrape cache AND the
+	// labelset is for a new series or the metadata for this series has just
+	// changed. It returns a boolean based on whether the metadata was updated.
+	updateMetadata := func(lset labels.Labels, isNewSeries bool) bool {
+		if !*conf.EnableMetadataStorage {
+			return false
+		}
+
+		cache.metaMtx.Lock()
+		defer cache.metaMtx.Unlock()
+		metaEntry, metaOk := cache.metadata[lset.Get(labels.MetricName)]
+		if metaOk && (isNewSeries || metaEntry.lastIterChange == cache.iter) {
+			metadataChanged = true
+			meta.Type = metaEntry.Type
+			meta.Unit = metaEntry.Unit
+			meta.Help = metaEntry.Help
+			return true
+		}
+		return false
+	}
+
+	// Take an appender with limits.
+	app = appender(app, *conf.SampleLimit, *conf.BucketLimit)
+
+	defer func() {
+		if err != nil {
+			return
+		}
+		// Only perform cache cleaning if the scrape was not empty.
+		// An empty scrape (usually) is used to indicate a failed scrape.
+		cache.iterDone(len(b) > 0)
+	}()
+
+loop:
+	for {
+		var (
+			et              textparse.Entry
+			isHistogram     bool
+			met             []byte
```

Comment:
Can we name to something longer and better? This is used far down in the loop. I can't understand what `met` is.